python/python/lance/dataset.py

```diff
@@ -641,16 +642,18 @@ def from_fragments(fragments: Iterable[LanceFragment], **kwargs):
         -------
         LanceScanner
         """
-        # TODO: peek instead of list
-        if not isinstance(fragments, list):
-            fragments = list(iter(fragments))
-
-        if len(fragments) == 0:
+        fragments_iter = iter(fragments)
+        try:
+            first = next(fragments_iter)
+        except StopIteration:
             raise ValueError("Must pass at least one fragment")
-        
-        dataset = fragments[0]._ds
-        # TODO: pass down an iterator
-        scanner = _Scanner.from_fragments([fragment._fragment for fragment in fragments])
+
+        # put first fragment back into iterator after we have peeked at it
+        dataset = first._ds
+        fragments_iter = itertools.chain([first], fragments_iter)
```

Comment:
This is a bit over-complicated, is this a common patten for `peekable iterator` in Python?

@changhiskhan what do you think.

Reply:
If it is hard to implement peekable iterator in python, maybe just let rust handle it? and throw an `PyValueError` from `pyro3` code?

Reply:
Well the reason we need it is so we can get the dataset out. What do you think of having the user pass in the dataset themselves? Then we wont need to peek and we can allow building scanners with zero fragments.

Reply:
If we change the signature of this to:

```python
@staticmethod
def from_fragments(fragments: Iterable[LanceFragment], dataset: LanceDataset, **kwargs):
   ...
```

Then we don't need to peek into the iterator. We need the dataset in both the call to `_Scanner.from_fragments()` and `LanceScanner()`. That's why the peek is happening in Python right now.